### ISSUE TITLE
V4 Development Tracker

### ISSUE THREAD

**Issue by normj** [aws/aws-sdk-net#3362]
### Describe the issue

Placeholder

### Links

placeholder

**Comment by dscpinheiro** [comment]
We'll publish the preview packages to NuGet (e.g. `<PackageReference Include="AWSSDK.AccessAnalyzer" Version="4.0.0.0-preview" />`), we're just working on some changes to our internal build system at the moment (we'll update this issue as soon as the packages are available).

**Comment by normj** [comment]
First preview packages have been released. Blog post for the release https://aws.amazon.com/blogs/developer/preview-1-of-aws-sdk-for-net-v4/

**Comment by martincostello** [comment]
Are you able to provide a preview version of `Amazon.AspNetCore.DataProtection.SSM` that depends on `AWSSDK.Extensions.NETCore.Setup`?

Trying to enable native AoT for a project that depends on the SSM package gives a bunch of warnings:

```text
ILC : Trim analysis error IL2026: Microsoft.Extensions.DependencyInjection.ExtensionMethods.PersistKeysToAWSSystemsManager(IDataProtectionBuilder,String,Action`1<PersistOptions>): Using member 'Microsoft.Extensions.DependencyInjection.ServiceCollectionExtensions.TryAddAWSService<IAmazonSimpleSystemsManagement>(IServiceCollection,ServiceLifetime)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. The AWSSDK.Extensions.NETCore.Setup package has not been updated to support Native AOT compilations. 
C:\codebuild\tmp\output\src1982900738\src\aws-sdk-net\sdk\src\Core\Amazon.Runtime\ConstantClass.cs(84): Trim analysis error IL2072: Amazon.Runtime.ConstantClass.Intern(): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields' in call to 'Amazon.Runtime.ConstantClass.LoadFields(Type)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
  ILC: Method '[AWSSDK.Extensions.NETCore.Setup]Amazon.Extensions.NETCore.Setup.DefaultClientConfig..ctor()' will always throw because: Missing method 'Void Amazon.Runtime.ClientConfig..ctor()'
  ILC: Method '[Amazon.AspNetCore.DataProtection.SSM]Amazon.AspNetCore.DataProtection.SSM.SSMXmlRepository+<GetAllElementsAsync>d__9.MoveNext()' will always throw because: Missing method 'Void Amazon.SimpleSystemsManagement.Model.GetParametersByPathRequest.set_WithDecryption(Boolean)'
```

It's possible some of the above are also coming from internal libraries for transient dependencies that haven't been compiled against v4 yet either.

The app is still blocked from 0 AoT warnings due to a dependency on another third-party library that hasn't published some fixes I made yet, but otherwise this SSM package is the final hurdle.

**Comment by martincostello** [comment]
FYI this is the error I get from an .NET 8 app running as a Lambda with the latest version of `Amazon.AspNetCore.DataProtection.SSM`. It builds fine, but then fails at runtime:

```text
System.Security.Cryptography.CryptographicException: An error occurred while trying to encrypt the provided data. Refer to the inner exception for more information.
 ---> System.MissingMethodException: Method not found: 'Void Amazon.SimpleSystemsManagement.Model.GetParametersByPathRequest.set_WithDecryption(Boolean)'.
   at Amazon.AspNetCore.DataProtection.SSM.SSMXmlRepository.GetAllElementsAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Amazon.AspNetCore.DataProtection.SSM.SSMXmlRepository.GetAllElementsAsync()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location ---
   at Amazon.AspNetCore.DataProtection.SSM.SSMXmlRepository.GetAllElements()
   at Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager.GetAllKeys()
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.CreateCacheableKeyRingCore(DateTimeOffset now, IKey keyJustAdded)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingProvider.GetCurrentKeyRingCore(DateTime utcNow, Boolean forceRefresh)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.Protect(Byte[] plaintext)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.Protect(Byte[] plaintext)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgeryTokenSerializer.Serialize(AntiforgeryToken token)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery.Serialize(IAntiforgeryFeature antiforgeryFeature)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery.GetAndStoreTokens(HttpContext httpContext)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.AntiforgeryExtensions.GetHtml(IAntiforgery antiforgery, HttpContext httpContext)
   at Microsoft.AspNetCore.Mvc.TagHelpers.FormTagHelper.Process(TagHelperContext context, TagHelperOutput output)
   at Microsoft.AspNetCore.Razor.TagHelpers.TagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)
   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.RunAsync(TagHelperExecutionContext executionContext)
   at AspNetCoreGeneratedDocument.Pages_Shared_SignIn.ExecuteAsync() in /_/src/Codeflowbot/Pages/Shared/SignIn.cshtml:line 13
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeResultAsync>g__Logged|22_0(ResourceInvoker invoker, IActionResult result)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|30_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at MyApp.MyLoggerMiddleware.Invoke(HttpContext context) in /_/src/Codeflowbot/Middleware/CurrentUserLoggerMiddleware.cs:line 20
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.InvokeCore(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddlewareImpl.<Invoke>g__Awaited|10_0(ExceptionHandlerMiddlewareImpl middleware, HttpContext context, Task task)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddlewareImpl.HandleException(HttpContext context, ExceptionDispatchInfo edi)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddlewareImpl.<Invoke>g__Awaited|10_0(ExceptionHandlerMiddlewareImpl middleware, HttpContext context, Task task)
   at MyApp.MyMiddleware.Invoke(HttpContext context) in /_/src/Codeflowbot/Middleware/LambdaLoggerMiddleware.cs:line 38
   at Amazon.Lambda.AspNetCoreServer.AbstractAspNetCoreFunction`2.ProcessRequest(ILambdaContext lambdaContext, Object context, InvokeFeatures features, Boolean rethrowUnhandledError)
```

**Comment by normj** [comment]
@martincostello A preview version of `Amazon.AspNetCore.DataProtection.SSM` is out targeting V4 https://www.nuget.org/packages/Amazon.AspNetCore.DataProtection.SSM/4.0.0-preview.1

**Comment by martincostello** [comment]
Thanks @normj.

I've tested the latest version of the packages with three of our applications collectively using the v4 packages listed below, and I've not encountered any issues:

- Amazon.AspNetCore.DataProtection.SSM
- AWSSDK.DynamoDBv2
- AWSSDK.SimpleSystemsManagement
- AWSSDK.SecurityToken

I've also tried native AoT publishing with two of the three, and there were also no trim warnings.

**Comment by normj** [comment]
Awesome @martincostello! Thanks for the feedback.

**Comment by Dreamescaper** [comment]
Would it make sense to drop 'netcoreapp3.1' target framework as well in V4? It is not supported by Microsoft. Anyone who uses it for whatever reason - would still be able to use netstandard2.0 version.

**Comment by normj** [comment]
@Dreamescaper When we choose to drop a target we look at is it supported by Microsoft and our customer usage of the target. For example .NET 6 goes out of support by Microsoft tomorrow but we know there are still a lot of users still using .NET 6 and that would pretty rough to force everybody to update to get access to the latest AWS changes.

I can't give numbers but when we started the V4 effort last spring there was still a lot of .NET Core 3.1 usage. The usage is dropping pretty fast now but given the amount of required heads up time we have to give for dropping a platform it would delay V4. I suspect we will do a follow up 4.1 sometime in the near future that drops .NET Core 3.1. There are also some slight differences between the .NET Core 3.1 target and .NET Standard 2.0 target with dependencies and a few APIs that are only available in .NET Core 3.1 given it has access to more .NET APIs.

**Comment by SamuelCox** [comment]
Is there any rough timeline on a preview 2 or GA? Excited to use this and see the perf benefits.

**Comment by normj** [comment]
@SamuelCox We are currently on preview 4. I'm hesitant to put a timeframe because as a team that is involved with almost every AWS release you can imagine we get randomized quite a bit. The intention was for V4 to have a relatively short development cycle to get to GA avoid it getting delayed due to our team's nature of being pulled in so many directions. Given we are in re:Invent time and then moving into the holiday season here in the US I don't see us GAing this year.

**Comment by juanchovelezpro** [comment]
is there an ETA when this will be GA?

**Comment by normj** [comment]
Preview 6 was released today. Big notable changes in this release are Native AOT support the DynamoDB high level libraries. That includes the DocumentModel and DataModel also known as Object Persistence Model libraries. For JSON based services the SDK switched to System.Text.Json for handling the JSON parsing. This change reduces allocations and garbage collection caused by serialization and deserialization for the AWS operations.

As for ETA sorry I can't give one. I'm hopeful we are getting close but as I warned above we are team that gets randomized quite a bit. We have a few stretch things we want to do but for the most part we have the feature changes we plan on complete. But we need to go through an internal review and some more protocol test validations and other miscellaneous internal processes for getting V4 to GA.

**Comment by normj** [comment]
Preview 7 was released today (February 17th, 2025). This preview was largely driven to unblock some PowerShell build issues taking in the V4 SDK. The release does contain more removal of LitJson although preview 6 contained the critical path changes to System.Text.Json. There are also a few breaking changes in areas we expect little usage. Checkout the changelog for full details: https://github.com/aws/aws-sdk-net/blob/v4-release/changelogs/SDK.CHANGELOG.2025.md#4000-preview7-2025-02-17-1610-utc

**Comment by normj** [comment]
Preview 8 released today (February 18th, 2025). Checkout the changelog for full details https://github.com/aws/aws-sdk-net/blob/v4-release/changelogs/SDK.CHANGELOG.2025.md#4000-preview8-2025-02-25-1338-utc. This preview does not contain a sync of the current V3 so does not have the latest service changes.

The main driver of the preview is this release has the final removal of the embedded LitJson parser. Making the core component a little smaller. The SDK is fully using System.Text.Json for all JSON parsing use cases. We are working on putting out new previews of our other high level libraries like [Amazon.AspNetCore.DataProtection.SSM](https://www.nuget.org/packages/Amazon.AspNetCore.DataProtection.SSM/) to target the new preview.

**Comment by Dreamescaper** [comment]
For DynamoDB client - do you think it would make sense to make `ConversionSchema.V1` as default one in V4?

**Comment by normj** [comment]
Preview 9 released today (March 4th, 2025). Change log for full details: https://github.com/aws/aws-sdk-net-staging/blob/v4-release/changelogs/SDK.CHANGELOG.2025.md#4000-preview9-2025-03-04-1429-utc

The main changes in preview 9 was a rework of the internal logic the SDK uses for resolving credentials. The motivators for this work were:
* Consistency with other AWS SDKs in particular the order in which credentials were search for.
* Flexibility at a per operation level what credentials and signer to use.

For consumers the what will be noticed is the `FallbackCredentialsFactory` type has been marked as `obsolete` and should be replaced with `DefaultAWSCredentialsIdentityResolver`. Also credential resolution will happen at the operation level instead as part of the service client constructor. That means if no credentials are available for the SDK the exception will happen during the operation call not service client construction.

The other significant change in preview 9 is how to configure output the SDK's logs.  Review the `Logging` section in starting comment thread for details. At a high level integrating to SDK logging with log frameworks has been moved to adaptor packages and there is not support for connecting the SDKs logging to Microsoft.Extensions.Logging.

**Comment by martincostello** [comment]
I've just updated an internal library we use `AWSSDK.SecurityToken` in to `4.0.0-preview.9` and we're getting a `NullReferenceException` when calling `AWSCredentials.GetCredentialsAsync()`:

```text
Unhandled exception. Amazon.Runtime.AmazonClientException: Error calling AssumeRole for role arn:aws:iam::<Account ID>:role/<Role Name>
---> System.NullReferenceException: Object reference not set to an instance of an object.
at Amazon.Runtime.Internal.Auth.AWS4Signer.Sign(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, BaseIdentity identity)
at Amazon.Runtime.Internal.Auth.AbstractAWSSigner.SignAsync(IRequest request, IClientConfig clientConfig, RequestMetrics metrics, BaseIdentity identity, CancellationToken token)
at Amazon.Runtime.Internal.Signer.SignRequestAsync(IRequestContext requestContext)
at Amazon.Runtime.Internal.Signer.PreInvokeAsync(IExecutionContext executionContext)
at Amazon.Runtime.Internal.Signer.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.BaseAuthResolverHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.Runtime.Internal.MetricsHandler.InvokeAsync[T](IExecutionContext executionContext)
at Amazon.SecurityToken.AmazonSecurityTokenServiceClient.Amazon.Runtime.SharedInterfaces.ICoreAmazonSTS_WebIdentity.CredentialsFromAssumeRoleWithWebIdentityAuthenticationAsync(String webIdentityToken, String roleArn, String roleSessionName, AssumeRoleWithWebIdentityCredentialsOptions options)
--- End of inner exception stack trace ---
at Amazon.SecurityToken.AmazonSecurityTokenServiceClient.Amazon.Runtime.SharedInterfaces.ICoreAmazonSTS_WebIdentity.CredentialsFromAssumeRoleWithWebIdentityAuthenticationAsync(String webIdentityToken, String roleArn, String roleSessionName, AssumeRoleWithWebIdentityCredentialsOptions options)
at Amazon.Runtime.AssumeRoleWithWebIdentityCredentials.GenerateNewCredentialsAsync()
at Amazon.Runtime.RefreshingAWSCredentials.<GetCredentialsAsync>g__GenerateCredentialsAndUpdateStateAsync|12_0()
at Amazon.Runtime.RefreshingAWSCredentials.GetCredentialsAsync()
<snipped>
```

**Comment by dscpinheiro** [comment]
@martincostello I don't think it's related to the `FallbackCredentialsFactory` but yes we also changed the signer in preview 9. 

I see it's an internal library, but could you share a short example on how you're using the `AWS4Signer`? We've updated all the internal usages but likely missed something.

**Comment by martincostello** [comment]
Sure, I'll put something together from it.

**Comment by martincostello** [comment]
The code is effectively what's shown below (I've amended it slightly to hardcode the credential delegate and service config that would come from the consuming code), and the exception is coming from an .NET container application running in EKS.

```csharp
using System.Text.Json;
using Amazon.Runtime;
using Amazon.Runtime.Credentials;
using Amazon.Runtime.Internal;
using Amazon.Runtime.Internal.Auth;
using Amazon.Runtime.Internal.Util;
using Amazon.SecurityToken;
using Amazon.SecurityToken.Internal;
using Amazon.SecurityToken.Model;
using Amazon.SecurityToken.Model.Internal.MarshallTransformations;

namespace IssueRepro;

public class IamHeaderRequestGenerator
{
    private static readonly AmazonSecurityTokenServiceEndpointResolver EndpointResolver = new();
    private readonly Func<AWSCredentials> _getAwsCredentials;
    private readonly AmazonSecurityTokenServiceConfig _amazonSecurityTokenServiceConfig;
    private readonly AWS4Signer _signer;

    public IamHeaderRequestGenerator()
    {
        _getAwsCredentials = DefaultAWSCredentialsIdentityResolver.GetCredentials;
        _amazonSecurityTokenServiceConfig = new() { RegionEndpoint = Amazon.RegionEndpoint.USEast1 };
        _signer = new AWS4Signer();
    }

    /// <inheritdoc/>
    public async Task<string> GenerateAsync(CancellationToken cancellationToken)
    {
        var identity = _getAwsCredentials();
        var credentials = await identity.GetCredentialsAsync().ConfigureAwait(false);

        var iamRequest = CreateRequest(credentials);

        Sign(iamRequest, identity);

        return EncodedHeaders(iamRequest);
    }

    private static string EncodedHeaders(IRequest iamRequest)
        => Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonSerializer.Serialize(iamRequest.Headers)));

    private void Sign(IRequest iamRequest, AWSCredentials identity)
    {
        _signer.Sign(
            iamRequest,
            _amazonSecurityTokenServiceConfig,
            new RequestMetrics(),
            identity);
    }

    private IRequest CreateRequest(ImmutableCredentials credentials)
    {
        var publicRequest = new GetCallerIdentityRequest();

        var iamRequest = GetCallerIdentityRequestMarshaller.Instance.Marshall(publicRequest);

        iamRequest.Endpoint = DetermineEndpointUrl(publicRequest);
        iamRequest.Headers["host"] = iamRequest.Endpoint.Host;

        iamRequest.ResourcePath = "/";

        if (!string.IsNullOrWhiteSpace(credentials.Token))
        {
            iamRequest.Headers.Add("X-Amz-Security-Token", credentials.Token);
        }

        iamRequest.Headers.Add("Content-Type", "application/x-www-form-urlencoded; charset=utf-8");

        return iamRequest;
    }

    private Uri DetermineEndpointUrl(AmazonWebServiceRequest request)
    {
        if (bool.TryParse(Environment.GetEnvironmentVariable("VAULT_USE_REGIONAL_STS_ENDPOINT"), out bool useRegionalStsEndpoint) && useRegionalStsEndpoint)
        {
            var requestContext = new RequestContext(false, _signer)
            {
                ClientConfig = _amazonSecurityTokenServiceConfig,
                Request = new DefaultRequest(request, "sts"),
            };

            var executionContext = new Amazon.Runtime.Internal.ExecutionContext(requestContext, new ResponseContext());

            var endpoint = EndpointResolver.GetEndpoint(executionContext);

            return new(endpoint.URL);
        }

        return new("https://sts.amazonaws.com");
    }
}
```

**Comment by martincostello** [comment]
I did try changing the above to something more like the code before preview 9, but that didn't work either:

```diff
- private void Sign(IRequest iamRequest, AWSCredentials identity)
+ private void Sign(IRequest iamRequest, ImmutableCredentials credentials)
{
+   var identity = new BasicAWSCredentials(credentials.AccessKey, credentials.SecretKey);
```

**Comment by normj** [comment]
@dscpinheiro has a fix going through PR right now. I'm curious @martincostello what is the use case are you solving with this code. You are relying on a lot of SDK code that we consider internal so I'm wondering what high level functionality we need that would avoid you taking dependency on parts we consider internal. Is this a demand for us having a general/public SigV4 signer?

For background we rely on the pattern of having "internal" namespaces to share code between service packages. We can't rely on the `InternalsVisibleTo` attribute because these methods have to be sharable to service packages that aren't known at the time core's creation.

**Comment by martincostello** [comment]
I don't know the historical details of _why_ the code does exactly what it does, but at a high-level the IAM role the application has is being used to authenticate API calls to an instance of Hashicorp Vault to retrieve secrets stored there.

If there's an easier/simpler way to do the same thing, then I'm all ears.

**Comment by dscpinheiro** [comment]
@martincostello We just released another preview (https://github.com/aws/aws-sdk-net/releases/tag/4.0.0.0-preview.10) with the PR Norm mentioned. Would you be able to verify it? (and thanks again for trying out the previews!)

**Comment by martincostello** [comment]
Sure - I'll give it a spin on Monday morning.

**Comment by martincostello** [comment]
To be annoying though (😅), I just found a _different_ issue trying preview.10 on a different personal project. My hunch is that it's because `AWSSDK.SecretsManager.Caching` doesn't have a preview built on top of the v4 SDK.

When I open https://github.com/martincostello/alexa-london-travel/commit/aab97bd10733afd84940aadb8f3825781c613d9f in Visual Studio, I get this error on build:

```text
An instance of analyzer Amazon.Lambda.CodeAnalysis.NullCollectionInitializerAnalyzer cannot be created from D:\packages\nuget\awssdk.lambda\4.0.0-preview.10\analyzers\dotnet\cs\AWSSDK.Lambda.CodeAnalysis.dll : Could not load type 'Amazon.CodeAnalysis.Shared.AbstractNullCollectionInitializerAnalyzer' from assembly 'SharedAnalysisCode, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'..
```

If it's not that, then maybe it's related to the Aspire things for the same reason.

**Comment by dscpinheiro** [comment]
Oh, that's a package that the SecretsManager team owns. 

We'll need to reach out to them to update to V4 too, thanks for catching that! FYI @normj

**Comment by normj** [comment]
@mbp The goal with the SDK is not receive any trim warnings. Can you `<TrimmerSingleWarn>false</TrimmerSingleWarn>` to your project file so we can see what in the SDK is triggering the warning. If you can share any repro code that would really help to.

**Comment by mbp** [comment]
@normj 

We are not able to identify the code from our code base that reproduces the issue. But here are the warnings in details.

```
aws-sdk-net\sdk\src\Services\DynamoDBv2\Custom\DataModel\ContextInternal.cs(552,21): Trim analysis error IL2067: Amazon.DynamoDBv2.DataModel.DynamoDBContext.FromDynamoDBEntry(SimplePropertyStorage, DynamoDBEntry, DynamoDBFlatConfig): 'targetType' argument does not satisfy 'DynamicallyAccessedMemberTypes.All' in call to 'Amazon.DynamoDBv2.DataModel.DynamoDBContext.DeserializeFromDocument(Document, Type, DynamoDBFlatConfig)'. The parameter 'value' of method 'System.Collections.Generic.Dictionary<TKey,TValue>.TryGetValue(Dictionary<TKey,TValue>.TKey, Dictionary<TKey,TValue>.TValue&)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

aws-sdk-net\sdk\src\Services\DynamoDBv2\Custom\DataModel\ContextInternal.cs(556,17): Trim analysis error IL2067: Amazon.DynamoDBv2.DataModel.DynamoDBContext.FromDynamoDBEntry(SimplePropertyStorage, DynamoDBEntry, DynamoDBFlatConfig): 'targetType' argument does not satisfy 'DynamicallyAccessedMemberTypes.All' in call to 'Amazon.DynamoDBv2.DataModel.DynamoDBContext.TryFromList(Type, DynamoDBList, DynamoDBFlatConfig, Dictionary<String,Type>, Object&)'. The parameter 'value' of method 'System.Collections.Generic.Dictionary<TKey,TValue>.TryGetValue(Dictionary<TKey,TValue>.TKey, Dictionary<TKey,TValue>.TValue&)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

aws-sdk-net\sdk\src\Services\DynamoDBv2\Custom\DataModel\ContextInternal.cs(710,17): Trim analysis error IL2072: Amazon.DynamoDBv2.DataModel.DynamoDBContext.ToDynamoDBEntry(SimplePropertyStorage, Object, DynamoDBFlatConfig, Boolean): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.All' in call to 'Amazon.DynamoDBv2.DataModel.DynamoDBContext.TryToMap(Object, Type, DynamoDBFlatConfig, Dictionary<Type,String>, Document&)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

aws-sdk-net\sdk\src\Services\DynamoDBv2\Custom\DataModel\ContextInternal.cs(714,17): Trim analysis error IL2072: Amazon.DynamoDBv2.DataModel.DynamoDBContext.ToDynamoDBEntry(SimplePropertyStorage, Object, DynamoDBFlatConfig, Boolean): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.Interfaces' in call to 'Amazon.DynamoDBv2.DataModel.DynamoDBContext.TryToList(Object, Type, DynamoDBFlatConfig, Dictionary<Type,String>, DynamoDBList&)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

aws-sdk-net\sdk\src\Services\DynamoDBv2\Custom\DataModel\ContextInternal.cs(717,17): Trim analysis error IL2072: Amazon.DynamoDBv2.DataModel.DynamoDBContext.ToDynamoDBEntry(SimplePropertyStorage, Object, DynamoDBFlatConfig, Boolean): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.All' in call to 'Amazon.DynamoDBv2.DataModel.DynamoDBContext.SerializeToDocument(Object, Type, DynamoDBFlatConfig, String)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

**Comment by normj** [comment]
@mbp Sorry it took me some time to get back to look into the trim warnings but I have put a PR to address the issue. https://github.com/aws/aws-sdk-net/pull/3749

**Comment by normj** [comment]
Updated the main description for preview 13 and status of V4.

**Comment by antoniofreire** [comment]
Hello.

Sorry if this isn't the right place to report, but i'm receiving a 414 response from AWS when trying to receive messages from a SQS queue using the 4.0.0-preview.13 version.

I've added a log to the request message and saw that, on each iteration, the User-Agent header is growing. It appears that the SDK is appending the following string to the header:

> aws-sdk-dotnet-coreclr/4.0.0.0 ua/2.1 os/linux#6.1.130.139 md/ARCH#Arm64 lang/.NET_Core#9.0.4 md/aws-sdk-dotnet-core#4.0.0.0 api/SQS#4.0.0.0 md/ClientAsync cfg/init-coll#0 m/b,P,0,E

After the first requet, the SDK duplicates the portion `aws-sdk-dotnet-coreclr/4.0.0.0 ua/2.1 os/linux#6.1.130.139 md/ARCH#Arm64 lang/.NET_Core#9.0.4 md/aws-sdk-dotnet-core#4.0.0.0 api/SQS#4.0.0.0 md/ClientAsync cfg/init-coll#0` of the User-Agent

**Comment by dscpinheiro** [comment]
> Hello.
> 
> Sorry if this isn't the right place to report, but i'm receiving a 414 response from AWS when trying to receive messages from a SQS queue using the 4.0.0-preview.13 version.
> 
> I've added a log to the request message and saw that, on each iteration, the User-Agent header is growing. It appears that the SDK is appending the following string to the header:
> 
> > aws-sdk-dotnet-coreclr/4.0.0.0 ua/2.1 os/linux#6.1.130.139 md/ARCH#Arm64 lang/.NET_Core#9.0.4 md/aws-sdk-dotnet-core#4.0.0.0 api/SQS#4.0.0.0 md/ClientAsync cfg/init-coll#0 m/b,P,0,E
> 
> After the first requet, the SDK duplicates the portion `aws-sdk-dotnet-coreclr/4.0.0.0 ua/2.1 os/linux#6.1.130.139 md/ARCH#Arm64 lang/.NET_Core#9.0.4 md/aws-sdk-dotnet-core#4.0.0.0 api/SQS#4.0.0.0 md/ClientAsync cfg/init-coll#0` of the User-Agent

We did release an user-agent related change as part of preview 13 too. Please create a separate issue and we'll take a look (I'm interested in the _trying to receive messages_ part - are you just calling the API in a loop or are you doing something else? Reproduction code will be needed for the investigation).

**Comment by normj** [comment]
Added an to the main description about some branch changes in preparation for V4 GA.

**Comment by normj** [comment]
Updated description for the GA announcement of V4. Thank you all for your support and feedback as developed V4.

**Comment by normj** [comment]
For anybody that noticed the [AWSSDK.Extensions.Bedrock.MEAI](https://www.nuget.org/packages/AWSSDK.Extensions.Bedrock.MEAI/) package mistakenly got released without the preview suffix. Since it has a dependency on [Microsoft.Extensions.AI.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) which is still in preview then `AWSSDK.Extensions.Bedrock.MEAI` should still be preview even if the SDK went GA. We have unlisted the 4.0.0 version and a `4.0.0-preview.14` version will be released soon that will target the GA version of the SDK.

**Comment by martincostello** [comment]
Just bumping this old comment as it doesn't have a v4 version yet: https://github.com/aws/aws-sdk-net/issues/3362#issuecomment-2707054425

**Comment by martincostello** [comment]
Does [Aspire.Hosting.AWS](https://github.com/aws/integrations-on-dotnet-aspire-for-aws) also need updating to support v4?

**Comment by normj** [comment]
@martincostello planning on updating `Aspire.Hosting.AWS` to V4 soon as well. I was thinking we would bundle that with updating the Aspire dependencies to Aspire 9.2.

**Comment by SamWilliamsGS** [comment]
This may be a stupid question, but I can't see any information in the documentation about why this change was made and how to migrate existing code to accomodate:

> Properties using value types on classes used for making requests and responses have been changed to use nullable value types. Essentially properties of type bool are changed to bool?.

I understand the change, but what does a `null` mean for these properties? Was it always possible that these response properties could be `null` and now we're more accurately reflecting it in the SDK types? Or was there a change in AWS responses? Is there a recommended way to migrate existing code that relies on these bools always being a bool? 🙂 Thanks

**Comment by normj** [comment]
@SamWilliamsGS There are cases where a value for a value type property is optional from a response. For example in CloudWatch Logs the `LogGroup` has a `RetentionInDays` property modeled as a integer. Having a retention policy is optional for a log group and if not set then `null` is the right value for `RetentionInDays`. 

Another problematic situation in V3 is there are some sets of APIs where you call the GET operation to get the object, do some mutation and then send the same instance in on an update cases. There were a class of bugs we had in V3 when we did that round trip and this value types that default to their default value like `false` for boolean could get misunderstood on the update call because it was originally no value and now looks like an explicit set to `false`.

To add more context when the SDK generates a type for the service operations the type could be used for either or both request and responses. Even if it is only used in one direction there is no rule it won't change in the future and be used in both directions. So we debated about different behavior depending on request or response usage but that turned not to be possible.

**Comment by peter-nguyen-contemi** [comment]
Hi, thank you for the great update. I'm not sure if this is the right thread for the question. But it would be great if the AWS.Cryptography.EncryptionSDK package will be updated to AWSSDK v4 too.

**Comment by normj** [comment]
> But it would be great if the AWS.Cryptography.EncryptionSDK package will be updated to AWSSDK v4 too.

Agreed. That library isn't own by our team but I'm reaching out to them to figure out how to get the library updated.

**Comment by catester** [comment]
I understand the reason for change for value type properties like bool properties but now `List<T>` properties also come as null, so we need to check for null before foreach loops everywhere now:

```c#
if (response.CommonPrefixes != null)
    foreach (var prefix in response.CommonPrefixes)
    {
    }

if (response.S3Objects != null)
    foreach (var s3Object in response.S3Objects)
    {
    }
```

So get ready to have NullReferenceException crashes with your app when you update to v4.x.
Just documenting these finding for future surfers.

Other value type properties are easier to handle:

```c#
Size = s3Object.Size ?? 0;

DateModified = s3Object.LastModified ?? DateTime.MinValue;
```

**Comment by dscpinheiro** [comment]
@catester It's mentioned in the [announcement](https://aws.amazon.com/blogs/developer/general-availability-of-aws-sdk-for-net-v4-0/), but you can use the `Amazon.AWSConfigs.InitializeCollections` property to keep the same behavior as V3 if you don't want (or can't) update your code base to check for null collections:

>  When updating to V4, if you are unsure whether your application is compatible with this runtime behavior change, the static property called `Amazon.AWSConfigs.InitializeCollections` can be set to `true`. This changes the runtime behavior to be the same as in V3, which initializes collection properties to an empty collection. Setting this property to `true` can be a good first stage when updating to V4, but doing so will negate the performance improvements and your ability to determine if a collection property is set or empty by default.

**Comment by martincostello** [comment]
> Does [Aspire.Hosting.AWS](https://github.com/aws/integrations-on-dotnet-aspire-for-aws) also need updating to support v4?

Just to confirm, lack of support for this is blocking one of my projects from upgrading as it blocks my Aspire host tests.

**Comment by normj** [comment]
@martincostello We just pushed out version `9.2.0` of Aspire.Hosting.AWS that targets Aspire 9.2 and V4 of AWS SDK for .NET.

**Comment by martincostello** [comment]
@normj Thanks - successfully upgraded to v4 now 🚀

**Comment by maxbog** [comment]
> [@catester](https://github.com/catester) It's mentioned in the [announcement](https://aws.amazon.com/blogs/developer/general-availability-of-aws-sdk-for-net-v4-0/), but you can use the `Amazon.AWSConfigs.InitializeCollections` property to keep the same behavior as V3 if you don't want (or can't) update your code base to check for null collections:
> 
> > When updating to V4, if you are unsure whether your application is compatible with this runtime behavior change, the static property called `Amazon.AWSConfigs.InitializeCollections` can be set to `true`. This changes the runtime behavior to be the same as in V3, which initializes collection properties to an empty collection. Setting this property to `true` can be a good first stage when updating to V4, but doing so will negate the performance improvements and your ability to determine if a collection property is set or empty by default.

@dscpinheiro Hi, we just stumbled upon this problem and I understand the reasons behind possibly null fields and collections. However, I don't understand why the fields and return values are not marked as nullable. This is a major problem because the AWS SDK is asking us to check for null on values that are marked as not-null, which actually is a problem detected by IDEs and the suggested solution by the IDE is to remove the null check. 

What plans do you have to address this issue? I'm not complaining about breaking backwards compatibility (that is fine), but about working against the language and IDEs. Right now, using V4 SDK is a sure way to introduce runtime `NullReferenceExceptions` with the library actively working to make sure those errors are hard to detect.

**Comment by kemsky** [comment]
`AWSSDKUtils.UrlEncode` was modified to match  RFC 3986 but now it does encode colon `:`.  Either it is not correct or method xml doc must be updated.

**Comment by normj** [comment]
@maxbog We had lots of debates on whether to use .NET nullable reference types. The major problems we had with using it was we still need to support .NET Framework which has questionable support for using nullable reference types. The other challenge is with the design of the SDK and every POCO having get and set methods for the properties we would essentially have to mark everything as nullable which would create a swarm of compiler warnings/errors for everybody.

We did add some Roslyn analyzers in V4 to help detect some of the null collection use cases. I think we need to invest more effort into our Roslyn analyzers to help catch potential null exceptions at compile time.

**Comment by doomchild** [comment]
Adding support for nullable reference types to .NET Framework requires implementing a couple of internal attributes, and then it works flawlessly. There's even a [package](https://www.nuget.org/packages/Nullable/) that will add support for you. Not having nullablility annotations is incredibly frustrating with modern tooling that expects them.

At a base level, though, trading unnecessary allocations for unexpected and widespread null reference exceptions seems like a bad move. Other tools like Localstack don't emulate the "things not marked as nullable might be null" behavior, so testing becomes more error-prone (I just spent two hours searching for the source of an SQS `ReceiveMessage` NRE in our build server that didn't occur in the testing environment because of exactly this problem).

From the perspective of someone consuming the package, this is a bad change. Even if the collection came back as a properly-annotated `null`, what does that communicate over an empty collection? For singular properties, I can see that as a useful bit of info (that `bool` wasn't set on the response from the service, so I know there are three possibilities: the service said `true`, the service said `false`, or the service said nothing) but for, say, the list of messages I get back from SQS, what value is there in that distinction? Either there are messages or there aren't. It requires me to do null-checks or coalesces all over the place, but I don't see how I'm getting any value for that cost.

**Comment by maxbog** [comment]
@normj 
> The other challenge is with the design of the SDK and every POCO having get and set methods for the properties we would essentially have to mark everything as nullable which would create a swarm of compiler warnings/errors for everybody.

Isn't that a good thing? You made a breaking change in the library, and now the compiler can help and show exactly where the library consumer needs to change their code to make it safe again. I don't understand the reasoning behind this. Why is it better to hide the information on where the change in the libraries can break client code?

> We did add some Roslyn analyzers in V4 to help detect some of the null collection use cases. I think we need to invest more effort into our Roslyn analyzers to help catch potential null exceptions at compile time.

So you're basically reimplementing those parts of the C# compiler? Why not use the perfect tool you already have instead, which, I bet, has better capabilities of detecting null references.

  It's just sad that will all those advancements in null analysis in the C# compiler, there is a widely popular library that actively circumvents years of work put into null safety in .NET.
  
  I kindly ask you to reconsider this decision and add the nullable annotations where applicable. Why waste time put into the analyzers, when you already have a production-ready one in the form of the compiler? Why do you force everyone using the libraries to manually scan all of their code for instances where this problem might appear, instead of helping them with the compiler warnings and errors?

**Comment by jamesivie** [comment]
> [@normj](https://github.com/normj)
> 
> > The other challenge is with the design of the SDK and every POCO having get and set methods for the properties we would essentially have to mark everything as nullable which would create a swarm of compiler warnings/errors for everybody.
> 
> Isn't that a good thing? You made a breaking change in the library, and now the compiler can help and show exactly where the library consumer needs to change their code to make it safe again. I don't understand the reasoning behind this. Why is it better to hide the information on where the change in the libraries can break client code?
> 
> > We did add some Roslyn analyzers in V4 to help detect some of the null collection use cases. I think we need to invest more effort into our Roslyn analyzers to help catch potential null exceptions at compile time.
> 
> So you're basically reimplementing those parts of the C# compiler? Why not use the perfect tool you already have instead, which, I bet, has better capabilities of detecting null references.
> 
> It's just sad that will all those advancements in null analysis in the C# compiler, there is a widely popular library that actively circumvents years of work put into null safety in .NET.
> 
> I kindly ask you to reconsider this decision and add the nullable annotations where applicable. Why waste time put into the analyzers, when you already have a production-ready one in the form of the compiler? Why do you force everyone using the libraries to manually scan all of their code for instances where this problem might appear, instead of helping them with the compiler warnings and errors?

It's much worse than that.  SDK V4 actively rejects everything that the non-nullability movement stands for.  It's not just the reference types that are nullable, now EVERYTHING is nullable!  And rather than document when things might or might not be null, it looks like we're just expected to always check for nulls outside the SDK.

**Comment by Vake93** [comment]
Not adopting nullable reference type (NRT) annotations in the AWS SDK V4 is a missed opportunity. There is a level of developer experience I've come to expect from AWS, and this feels like a step backwards. Having APIs that still feel "pre-C# 8" in 2025 diminishes confidence in the SDK as a modern, first-class experience.

Some of the reasons given for not adopting NRTs seem questionable from my POV:

- ".NET Framework has questionable support for NRTs" - NRTs are a C# compiler feature introduced in C# 8. They don’t require runtime support. It's entirely possible to target .NET Framework with libraries that use NRTs.

- "We’d have to mark everything as nullable, causing a swarm of warnings" - If most properties can be null, then marking them as nullable is simply being accurate. The alternative misleads consumers into believing values are non-null when they aren't. Yes, this will generate warnings in consuming code, but that is the purpose of NRTs is to catch potential null issues at compile time rather than at runtime.

- "We’ll rely on Roslyn analyzers instead" - Custom analyzers are not a substitute for language native nullability. They cover only a subset of cases, add cognitive overhead. This forces developers to learn AWS-specific analyzer rules rather than using the standardized nullability system built into the C# compiler.

The AWS V4 SDK already introduces breaking changes. A major version update is the right time to modernize the API contracts and models with NRTs. It will improve developer experience and aligning the SDK with modern C# practices.

**Comment by hainguyenthanh5-oss** [comment]
Hi, will the [AWS.Cryptography.EncryptionSDK](https://www.nuget.org/packages/AWS.Cryptography.EncryptionSDK) package be updated to AWSSDK v4? This package has not received any updates for more than 1 year. If you know the people responsible for maintaining this package, please forward my question to them. Thank you in advance.

**Comment by normj** [comment]
@hainguyenthanh5-oss We have reached out to the team that owns that package. They plan on updating it but there are other logistical tasks they have to take care of first before they can get that package updated.

**Comment by gabor-bocso** [comment]
Should we expect getting de / serialization errors by changing from LitJson to System.Text.Json, because of missing converters?
I don't know LitJson (just checked and its not even on version 1.00...) I think this change alone should get more attention.

**Comment by peterrsongg** [comment]
> Should we expect getting de / serialization errors by changing from LitJson to System.Text.Json, because of missing converters? I don't know LitJson (just checked and its not even on version 1.00...) I think this change alone should get more attention.

@gabor-bocso can you elaborate? are you seeing any ser/de errors in particular?

**Comment by normj** [comment]
Resolving the issue now that V4 has been GA for awhile and all other libraries that our team own or have influence with have also been updated to V4

In response to the nullability comments. I updated the repo's [README](https://github.com/aws/aws-sdk-net?tab=readme-ov-file#null-collections-and-nullable-reference-types) with our explanation of the choices we made for nullability. I recognize the answer might frustrate some users and is an issue I constantly debate but at this point V4 is GA and we can't introduce reference types now as that would be a breaking change. We consider introducing all of the warnings reference types would trigger a breaking change. When we do a V5, which is not planned, will revisit this again. Keep in mind there are also users that don't like value types are using nullable in V4 which we had to for model correctness, so there isn't a right choice for all users unless we dramatically change the programming model. V4 was never going to change the programming model as it was scoped to be an evolutionary change.

When we do start a V5 I highly recommend getting involved in early previews. The nullable changes in V4 were made in preview 1 to collect feedback and we had 13 previews over the course of 7 months. I'm not attempting to shame anybody about preview feedback, we are all busy and it is hard to find time to look at previews, but hopefully it is clear we did try and be transparent with this change well before GA.
